openmc/mesh.py

```diff
@@ -624,6 +628,333 @@ def to_xml_element(self):
+        t_grid_str = str(self._phi_grid) if self._phi_grid is None else len(self._phi_grid)
+        string += fmt.format('\tN Phi pnts:', '=\t', t_grid_str)
+        if self._phi_grid is not None:
+            string += fmt.format('\tPhi Min:', '=\t', self._phi_grid[0])
+            string += fmt.format('\tPhi Max:', '=\t', self._phi_grid[-1])
+        p_grid_str = str(self._z_grid) if self._z_grid is None else len(self._z_grid)
+        string += fmt.format('\tN Z pnts:', '=\t', p_grid_str)
+        if self._z_grid is not None:
+            string += fmt.format('\tZ Min:', '=\t', self._z_grid[0])
+            string += fmt.format('\tZ Max:', '=\t', self._z_grid[-1])
+        return string
+
+    @classmethod
+    def from_hdf5(cls, group):
+        mesh_id = int(group.name.split('/')[-1].lstrip('mesh '))
+
+        # Read and assign mesh properties
+        mesh = cls(mesh_id)
+        mesh.r_grid = group['r_grid'][()]
+        mesh.phi_grid = 180 / np.pi * group['p_grid'][()]
+        mesh.z_grid = group['z_grid'][()]
+
+        return mesh
+
+    def to_xml_element(self):
+        """Return XML representation of the mesh
+
+        Returns
+        -------
+        element : xml.etree.ElementTree.Element
+            XML element containing mesh data
+
+        """
+
+        element = ET.Element("mesh")
+        element.set("id", str(self._id))
+        element.set("type", "cylindrical")
+
+        subelement = ET.SubElement(element, "r_grid")
+        subelement.text = ' '.join(map(str, self.r_grid))
+
+        subelement = ET.SubElement(element, "p_grid")
+        subelement.text = ' '.join(map(str, self.phi_grid))
+
+        subelement = ET.SubElement(element, "z_grid")
+        subelement.text = ' '.join(map(str, self.z_grid))
+
+        return element
+
+    def calc_mesh_volumes(self):
+        """Return Volumes for every mesh cell
+
+        Returns
+        -------
+        volumes : Iterable of float
+            Volumes
+
+        """
+
+        # V = int_r int_phi int_phi r dr dphi dz
+
+        V_r = np.diff(np.array(self.r_grid)**2 / 2)
+        V_p = np.diff(np.array(self.phi_grid) * np.pi / 180.0)
+        V_z = np.diff(np.array(self.z_grid))
+
+        return np.multiply.outer(np.outer(V_r, V_p), V_z)
+
+
+class SphericalMesh(MeshBase):
+    """A 3D spherical mesh
+
+    Parameters
+    ----------
+    mesh_id : int
+        Unique identifier for the mesh
+    name : str
+        Name of the mesh
+
+    Attributes
+    ----------
+    id : int
+        Unique identifier for the mesh
+    name : str
+        Name of the mesh
+    dimension : Iterable of int
+        The number of mesh cells in each direction.
+    n_dimension : int
+        Number of mesh dimensions (always 3 for a RectilinearMesh).
+    r_grid : Iterable of float
+        Mesh boundary points along the r-axis.
+    theta_grid : Iterable of float
+        Mesh boundary points along the theta-axis.
```

Comment:
It would be good to clarify that the values are expected to be in degrees for any of the relevant class attributes.